SYCL/ESIMD/api/functional/mutator.hpp

```diff
@@ -0,0 +1,115 @@
+//===-- mutator.hpp - This file provides common function and classes to mutate
+//      reference data. ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides common function and classes to mutate reference data.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "type_traits.hpp"
+#include "value.hpp"
+#include <sycl/sycl.hpp>
+
+// for std::for_each
+#include <algorithm>
+#include <vector>
+
+namespace esimd_test::api::functional {
+
+// In some test cases it’s possible to pass a reference data as the input data
+// directly without any modification. For example, if we check some copy
+// constructor or memory move operation it’s OK to have any data values as the
+// input ones. But in most cases we need to consider the possibility of UB for
+// C++ operations due to modification of input values.
+//
+// Mutation mechanism covers such requirement.
+// The mutator namespace is intended to store such mutators alongside with
+// the generic ones provided below.
+namespace mutator {
+
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `val` later.
+template <typename T> class For_subtraction {
+  T m_value;
+
+public:
+  For_subtraction(T val)
+      : m_value((assert(val > 0 && "Invalid value."), val)) {}
+
+  void operator()(T &val) {
+    // Floating point variable with infinity value shouldn't be changed.
+    if constexpr (type_traits::is_sycl_signed_v<T>) {
+      if constexpr (type_traits::is_sycl_floating_point_v<T>) {
+        // sycl::half will be converted to float
+        if (std::isinf(val)) {
+          return;
+        }
+      }
+
+      const T lower_border = value<T>::lowest() + m_value;
+      if (val < lower_border) {
+        val = lower_border;
+      }
```

Comment:
```suggestion
      const T lower_border = value<T>::lowest() + m_value;
      if (val < lower_border) {
        // Validate only the negative infinity case, as the NaN will not enter this branch
        if constexpr (type_traits::is_sycl_floating_point_v<T>) {
          // sycl::half will be converted to float
          if (std::isinf(val)) {
            return;
          }
        }
	// we need to update value to avoid UB during addition
        val = lower_border;
      }
```

Reply:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/789/commits/41df5e531f17caed6253334af281f97f021058d3#diff-fa18495d278b576095ad26f3fbce0799c62e2302a8837d4e0a289c922d0dce74R47-R64) and in other mutators